### PR TITLE
Implement PacketBuilderStep<IpHeader>.write

### LINKED ISSUE #26: IP packets can't be written without a transport protocol
Right now, packets cannot be written out unless they are fully-assembled transport-layer TCP/UDP packets. In other words, the `write` function is only implemented for `PacketBuilder<UdpPacket>` and `PacketBuilder<TcpPacket>`:

- [PacketBuilder\<UdpPacket\>::write](https://docs.rs/etherparse/latest/etherparse/struct.PacketBuilderStep.html#method.write)
- [PacketBuilder\<TcpPacket\>::write](https://docs.rs/etherparse/latest/etherparse/struct.PacketBuilderStep.html#method.write-1)

However, there might be use cases for `etherparse` where a programmer may want to write an IP packet with a custom transport protocol or handle the protocol assembly and writing themselves. I suggest implementing `write` for `PacketBuilder<IpPacket>` to support this use case!

### PR DESCRIPTION
**Pull request by karpawich**
Resolves #26

@JulianSchmid : what is missing from this PR? I can help write tests, docs, etc. but I want to ask you for your guidance first.

### CODE REVIEW

**Comment by JulianSchmid** [comment]
Hi @karpawich ,

I don't know when I get around to reviewing your PR in depth. But add some tests here https://github.com/JulianSchmid/etherparse/blob/master/tests/packet_builder/mod.rs  for the new code paths you added.

You can use the coverage report to find holes and use the script https://github.com/JulianSchmid/etherparse/tree/master/scripts to calculate a coverage report locally.

Greets & 1000 Thanks
Julian

**Comment by JulianSchmid** [comment]
I forgot to mention: Thanks for the PR :)

**Comment by karpawich** [comment]
I have written the tests. I still need to update the documentation.

N.B: The `protocol` argument of the `Ipv4Header::new` function needs to be of type `u8` instead of `IpNumber` to support the use of unassigned IP numbers.

**Comment by JulianSchmid** [comment]
Hi @karpawich , I will try to have a look at your PR this weekend.

**Review by JulianSchmid — CHANGES_REQUESTED** [review]
Some minor changes that I will commit in a second.

File: src/packet_builder.rs [hunk 1]
```diff
@@ -614,7 +614,7 @@ impl PacketBuilderStep<VlanHeader> {
     ///         Ipv4Header::new(
     ///             0, //payload_len will be replaced during write
     ///             12, //time_to_live
-    ///             IpNumber::Udp, //will be replaced during write
+    ///             IpNumber::Udp as u8, //will be replaced during write
```
Comment (JulianSchmid) [inline src/packet_builder.rs:617]:
```suggestion
    ///             ip_number::UDP, //will be replaced during write
```

File: src/packet_builder.rs [hunk 2]
```diff
@@ -1373,75 +1383,93 @@ fn final_write<T: io::Write + Sized, B>(builder: PacketBuilderStep<B>, writer: &
...
-                    Udp(_) => ip_number::UDP,
-                    Tcp(_) => ip_number::TCP
+    //transport header
+    let transport = builder.state.transport_header;
+    match transport {
+        None => {
+            match ip_header {
+                Version4(mut ip, _) => {
+                    ip.set_payload_len(payload.len())?;
+                    ip.write(writer)?;
```
Comment (JulianSchmid) [inline src/packet_builder.rs:1398]: Missing write of extension headers

File: src/packet_builder.rs [hunk 3]
```diff
@@ -1373,75 +1383,93 @@ fn final_write<T: io::Write + Sized, B>(builder: PacketBuilderStep<B>, writer: &
...
+                },
+                Version6(mut ip, _) => {
+                    ip.set_payload_length(payload.len())?;
+                    ip.write(writer)?;
```
Comment (JulianSchmid) [inline src/packet_builder.rs:1402]: Missing write of extension headers

File: src/packet_builder.rs [hunk 4]
```diff
@@ -1373,75 +1383,93 @@ fn final_write<T: io::Write + Sized, B>(builder: PacketBuilderStep<B>, writer: &
...
+    let transport = builder.state.transport_header;
+    match transport {
+        None => {
+            match ip_header {
+                Version4(mut ip, _) => {
+                    ip.set_payload_len(payload.len())?;
+                    ip.write(writer)?;
+                },
+                Version6(mut ip, _) => {
+                    ip.set_payload_length(payload.len())?;
```
Comment (JulianSchmid) [inline src/packet_builder.rs:1401]: length of the extension headers

File: src/packet_builder.rs [hunk 5]
```diff
@@ -1373,75 +1383,93 @@ fn final_write<T: io::Write + Sized, B>(builder: PacketBuilderStep<B>, writer: &
...
-                    Icmpv6(_) => ip_number::IPV6_ICMP,
-                    Udp(_) => ip_number::UDP,
-                    Tcp(_) => ip_number::TCP
+    //transport header
+    let transport = builder.state.transport_header;
+    match transport {
+        None => {
+            match ip_header {
+                Version4(mut ip, _) => {
+                    ip.set_payload_len(payload.len())?;
```
Comment (JulianSchmid) [inline src/packet_builder.rs:1397]: length of the extension headers

File: src/packet_builder.rs [hunk 6]
```diff
@@ -1200,6 +1200,16 @@ impl PacketBuilderStep<IpHeader> {
             _marker: marker::PhantomData::<TcpHeader>{}
         }
     }
+
+    ///Write all the headers and the payload.
+    pub fn write<T: io::Write + Sized>(self, writer: &mut T, payload: &[u8]) -> Result<(),WriteError> {
```
Comment (JulianSchmid) [inline src/packet_builder.rs:1205]: I would add the last ip_number as an argument here so it can be set in the last extension header or ip header.

**Review by JulianSchmid — CHANGES_REQUESTED** [review]
I did another review round. You don't have to do anything I already have a commit ready with fixes to address the comments.

**Review by JulianSchmid — APPROVED** [review]

**Pull request merged**